src/QuadForm/Herm/GenusRep.jl

```
@@ -48,86 +8,171 @@ end
-      neig = iterated_neighbours(L, P0, max = max, use_auto = use_auto, missing_mass = missing_mass)
-      append!(result, neig)
-      max = max - length(result)
+  local C::ideal_type(R)
+
+  if CC isa Int
+    C = split ? a(P) : P
+  else
+    C = CC
+  end
+
+  I = Int[ i for i in 1:n if !iszero(xG[i])]
+  i = I[1]
+  M = zero_matrix(K, n - length(I), ncols(B))
+  for (k, nk) in enumerate(setdiff(1:n, I))
+    for j in 1:ncols(B)
+      M[k, j] = B[nk, j]
     end
-    for i in 1:length(result)
-      for j in 1:i-1
-        @hassert :GenRep 1 !(isisometric(result[i], result[j])[1])
-      end
+  end
+  CI = fractional_ideal_type(R)[ K(1) * R for j in 1:(n - length(I)) ]
+  for j in I
+    if j != i
+      M = vcat(M, matrix(K, 1, ncols(B), elem_type(K)[B[j, k] - a(h\(divexact(xG[j], xG[i]))) * B[i, k] for k in 1:ncols(B)]))
+      push!(CI, K(1) * R)
     end
+  end
+  M = vcat(M, sub(B, i:i, 1:ncols(B)))
+  push!(CI, fractional_ideal(order(P), P))
+  M = vcat(M, matrix(K, 1, ncols(B), x))
+  push!(CI, inv(C))
+  pm = pseudo_matrix(M, CI)
+  _M = _sum_modules(L, _module_scale_ideal((split ? P : P * C), pseudo_matrix(L)), pm)
+  LL = lattice(ambient_space(L), _M)
+
+  @assert islocally_isometric(L, LL, P)
+  return LL
+end
+
+function stdcallback(list, L)
+  keep = all(LL -> !isisometric(LL,L)[1], list)
+  return keep, true
+end
+
+#TODO : implement isometry check for indefinite lattice in the same space
+function relcallback(list,L)
+  if isdefinite(L)
+    keep = all(LL -> !isisometric(LL,L)[1], list)
   else
-    result = LL
+    keep = all(LL -> LL!= L, list) #In the indefinite case, we can't do better than this for now
   end
-  return result
+  return keep, true
 end
 
 function _neighbours(L, P, result, max, callback = stdcallback, use_auto = true)
-  ok, scale = ismodular(L, P);
+  ok, scale = ismodular(L, P)
   @req ok "The lattice must be locally modular"
   R = base_ring(L)
   K = nf(R)
   a = involution(L)
+
+  if !isdefinite(L) && use_auto
```

I would rather raise an error instead of silently overwriting the user input.

---

One problem with that is that if we follow through with this everywhere, we could not change the default `use_auto` to `true` in `iterated_neighbors`? (Because then for indefinite lattices it would be per default an error). I would rather document it in the form: "is `use_auto` is `true`, then the function will use automorphisms in the definite case. In the indefinite case this keyword has no effect."

---

Yes, you are right.. Here I did this because, by default, `use_auto = true`, so I just adapted the automatic output to the indefinite case. Maybe it will more natural to set `use_auto = false` and then turn this into an input check... should I try to make this change ?

---

I didn't see Tommy's comment before posting mine. I have actually followed this and change the documentations by explicitly saying that `use_auto` won't have any effect if the lattice is indefinite, and the boolean value of this entry is considered only in the definite case.